File: Go/FindMinimumInRotatedArray/Solution_test.go
package solution

import "testing"
import "fmt"

func TestSolution(t *testing.T) {
	sol := new(Solution)

	fmt.Print("Test NULL: ")
	minVal, ok := sol.FindMin([]int{})
	if ok == nil {
		t.Fatalf("Empty array must return null")
	}
	fmt.Println("OK")

	fmt.Print("Test Duplicates: ")
	minVal, ok = sol.FindMin([]int{1, 3, 4})
	if minVal != 1 {
		t.Fatalf("Should get %v but get %v", 1, minVal)
	}

	fmt.Print("Test unrotated array: ")
	num1 := []int{1, 2, 3, 4, 5, 6, 7}
	minVal, _ = sol.FindMin(num1)
	if minVal != 1 {
		t.Fatalf("Should get %v but get %v", 1, minVal)
	}
	fmt.Println("OK")

	fmt.Print("Test rotated array: ")
	num2 := []int{4, 5, 6, 7, 1, 2, 3}
	minVal, _ = sol.FindMin(num2)
	if minVal != 1 {
		t.Fatalf("Should get %v but get %v", 1, minVal)
	}
	fmt.Println("OK")
}

File: Scala/src/FindMinimumInRotatedSortedArray_Dup/SolutionTest.scala
package FindMinimumInRotatedSortedArray_Dup

import org.scalatest.FunSuite

class SolutionTest extends FunSuite {
  val sol = new Solution
  test("Test single element") {
    val num1 = Array(0)
    assert(sol.findMin(num1) == 0)
  }
  test("Test unrotated array") {
    val num2 = Array(1, 2, 3, 4, 5, 6)
    assert(sol.findMin(num2) == 1)
  }
  test("Test rotated array") {
    val num3 = Array(3, 4, 5, 6, 1, 2)
    assert(sol.findMin(num3) == 1)
  }
  test("Test unrotated dup0") {
    val num4 = Array(1, 2, 3, 4, 4, 5, 5, 6)
    assert(sol.findMin(num4) == 1)
  }
  test("Test unrotated dup1") {
    val num5 = Array(1, 3, 3, 3, 3, 3, 3, 3)
    assert(sol.findMin(num5) == 1)
  }
  test("Test rotated dup0") {
    val num6 = Array(3, 3, 3, 1, 3, 3, 3, 3)
    assert(sol.findMin(num6) == 1)
  }
  test("Test rotated dup1") {
    val num6 = Array(2, 0, 1, 1, 1)
    assert(sol.findMin(num6) == 0)
  }
  test("Test rotated dup2") {
    val num6 = Array(2, 2, 2, 0, 2, 2)
    assert(sol.findMin(num6) == 0)
  }
}
